README.md

```diff
@@ -1,24 +1,278 @@
-# Struct
+# Struct [![Build Status](https://img.shields.io/travis/ExpressApp/struct.svg)](https://travis-ci.org/ExpressApp/struct) [![Hex.pm](https://img.shields.io/hexpm/v/struct.svg)](https://hex.pm/packages/struct)
 
-**TODO: Add description**
+---
+
+Library for dealing with data structures
+
+---
 
 ## Installation
 
-If [available in Hex](https://hex.pm/docs/publish), the package can be installed as:
+1. Add `struct` to your list of dependencies in `mix.exs`:
+
+```elixir
+def deps do
+  [{:struct, "~> 1.0.0"}]
+end
+```
+
+2. Ensure `struct` is started before your application:
+
+```elixir
+def application do
+  [applications: [:struct]]
+end
+```
+
+## Usage
+
+Suppose you have some user input from several sources (DB, HTTP request, WebSocket), and you will need to process that data into something type-validated, like User entity. With this library you can define a type-validated struct for this entity:
+
+```elixir
+defmodule User do
+  use Struct
+
+  structure do
+    field :name
+    field :age, :integer
+  end
+end
+```
+
+And use it to cast your data into something identical, to prevent type coercion in different places of your code. Like this:
+
+```elixir
+iex> User.make(%{"name" => "John Doe", "age" => "37"})
+{:ok, %User{age: 37, name: "John Doe"}}
+```
+
+Pretty neat, yeah? But what if you need more complex type? We have a solution!
+
+```elixir
+defmodule Answer do
+  @behaviour Struct.Type
+
+  def cast("yes"), do: {:ok, true}
+  def cast("no"), do: {:ok, false}
+  def cast(_), do: {:error, :invalid_answer}
+end
+```
+
+And use it in your struct like this:
+
+```elixir
+defmodule Quiz do
+  use Struct
+
+  structure do
+    field :user_id, :integer
+    field :answers, {:array, Answer}
+  end
+end
+```
+
+```elixir
+iex> Quiz.make(%{user_id: 42, answers: ["yes", "no", "no", "yes"]})
+{:ok, %Quiz{answers: [true, false, false, true], user_id: 42}}
+```
+
+What if we need to parse 'optimized' query string from URL? Like list of user_id separated by a comma? Create a custom type for each boxed type? No! Use type composition:
```

Comment:
What if we need to parse 'optimized' query string from URL, like list of user ids separated by a comma?
Do we need to create a custom type for each boxed type? No! Just use type composition feature: